### ISSUE TITLE
GC is required after destroying an instance of Instagram class

### ISSUE THREAD

**Issue by crewmango** [mgp25/Instagram-API#1861]
### Before submitting an issue make sure you have:
- [X] [Searched](https://github.com/mgp25/Instagram-API/search?type=Issues) the bugtracker for similar issues including **closed** ones
- [X] [Read the FAQ](https://github.com/mgp25/Instagram-API/wiki/FAQ)
- [X] [Read the wiki](https://github.com/mgp25/Instagram-API/wiki)
- [X] [Reviewed the examples](https://github.com/mgp25/Instagram-API/tree/master/examples)
- [X] [Installed the api using ``composer``](https://github.com/mgp25/Instagram-API#installation)

### Purpose of your issue?
- [x] Bug report (encountered problems/errors)
- [ ] Question
- [ ] Other
- [ ] Feature request (request for new functionality)

Example Code:

```php
set_time_limit(0);
date_default_timezone_set('UTC');

require __DIR__.'/../vendor/autoload.php';

echo memory_get_usage() . PHP_EOL; // 4109424
$ig = new \InstagramAPI\Instagram();
echo memory_get_usage() . PHP_EOL; // 4501880
unset($ig);
echo memory_get_usage() . PHP_EOL; // 4501888
```

Creating and unsetting multiple Instagram instances results in a memory leak. I tried unsetting the RequestCollection instances before destroying the main instance (since they hold a reference to the parent), same result.

**Comment by valga** [comment]
It's not a memory leak, it's circular references. Call `gc_collect_cycles()` after `unset()` and it will clear them.

However, it might become a problem, because garbage collecting is expensive in some cases. Quite unfortunately, fixing these circular references will require a lot of refactoring, so it can't be done at the moment.

**Comment by crewmango** [comment]
I already tried a lot of things, including `gc_collect_cycles()` after `unset()` but makes absolutely no difference, memory stays at the same as after creating the instance. 
The issue becomes obvious in a loop, like this:
```
set_time_limit(0);
date_default_timezone_set('UTC');

require __DIR__.'/../vendor/autoload.php';
$instances = [];
echo memory_get_usage() . PHP_EOL; // 4109424
for ($i = 0; $i < 10; $i++) {
    $instances[$i] = new \InstagramAPI\Instagram();
    echo memory_get_usage() . PHP_EOL; // 4501880 (increases with every iteration)
    unset($instances[$i]);
    gc_collect_cycles();
    echo memory_get_usage() . PHP_EOL; // 4501888 (increases with every iteration)
}
echo memory_get_usage() . PHP_EOL; // 4590840
```

I would like to help with this issue if you can point me in the right direction.

**Comment by valga** [comment]
@crewmango First of all, below is an output for you snippet without and with GC.
Without:
```
1150696
3013968
3013968
3023800
3023800
3033632
3033632
3043464
3043464
3053296
3053296
3063128
3063128
3072960
3072960
3082792
3082792
3092624
3092624
3102456
3102456
3102456
```
With:
```
1150696
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3013968
3004136
3004136
```
It is clearly visible that GC works.

Now let me explain the gap between first and last rows (1MB vs 3MB). It happens because PHP reads and stores classes in memory as it loads them. Consider following example:
```php
require __DIR__.'/vendor/autoload.php';

echo memory_get_usage() . PHP_EOL;
class_exists(\InstagramAPI\Instagram::class, true);
echo memory_get_usage() . PHP_EOL;
class_exists(\InstagramAPI\Request::class, true);
echo memory_get_usage() . PHP_EOL;
```
Its output:
```
1150744
1246520
1298136
```
And finally, circular references. We have a lot of them here: https://github.com/mgp25/Instagram-API/blob/233ce87890976c3cfd0d86f9e8c694d91127df92/src/Instagram.php You can see `$this->something = new Something($this);` pattern that must be refactored.

**Comment by aitte2** [comment]
@valga 

> You can see $this->something = new Something($this); pattern that must be refactored.

But how can that be changed? Every request must be able to call `request()` to create a Request on the exact parent Instagram instance to use its active cookies and settings.

Do you propose this? `$feed = \InstagramAPI\Request\Timeline::getUserFeed(Instagram $ig, $userPk, $maxId)`? That would suck :P

**Comment by valga** [comment]
@aitte2 You are going into a wrong direction. We have a typical God object that holds everything, so we have a bigger problem than cyclic references. But killing a God is hard, and I don't have that much time now.

But if you are interested, I can give you some ideas.

**Comment by aitte2** [comment]
Ok :) don't know if I have time either but if you explain it we can talk about it and come up with a plan.

**Comment by SteveJobzniak** [comment]
@valga Hey man long time no see! Hope you're doing well and enjoying spring! I am not back, just felt like checking in on the project and its status today and saw this thread. And I have to say that this problem isn't about God Objects. :-)

A God Object is an object that is full of too many functions/algorithms ("does too much work in a single class") and where its code (number of functions etc) becomes unmanageably large. That's how the library code was in the old days when all ~400 functions were in Instagram.php.

The "cure" for a God Object is to take code out of it and put it in grouped modules/classes instead (exactly as this library is arranged now). Such modular design is the opposite of a God Object.

When you do that, the functions that used to be in a single location are now spread out, and therefore the new classes may need to call each other or get data from a "parent container" and _will need_ a reference to the other classes or "their parent" since what was previously an accessible function on the same class is now placed in another class.

That's exactly what this library does now. Every module references their parent so that the modules can still get critical data from the parent _and_ reach other modules to use their functions when necessary. And that's not a problematic pattern itself. That's only a problem in **the PHP language** because **its** garbage collector is _very_ bad (I actually didn't know that PHP's garbage collector sucked this much). It uses basic/very naive reference counting rather than tracing, so it is super bad at discovering when objects are truly dead.

To solve this common modular need (being able to have circular references between modules and parents, but that can still be garbage collected), PHP is working on a solution:

https://wiki.php.net/rfc/weakreferences

It was voted in September 2017 and they said yes to add this feature: 6 - to - 5: https://wiki.php.net/rfc/weakreferences/vote

With it we would be able to say that the references between parent and modules are "weak" and allowed to be garbage collected. Which means PHP's naive/basic garbage collector will be able to collect them properly. But that feature will only arrive in something like PHP 7.4. So we can't use it.

Another solution is two unofficial PHP modules mentioned here: https://medium.com/@johann.pardanaud/about-circular-references-in-php-10f71f811e9, but most webhosts don't allow installing modules. So we can't use it.

And totally breaking the circular references in this library doesn't seem possible. Many of their functions do not return until they have executed many separate steps. They _need_ to call functions in other modules and do lots of work before they can return anything.

Therefore we can't make functions that just return a "template of the request" that the core class can fill with the missing data (uuid, cookies, etc) and then execute. Something like `$ig->request(InstagramAPI\Request\Media::edit())`. Because it wouldn't be able to do all necessary multisteps that must be done to achieve the full goal.

With chained promises and callbacks we _may_ be able to achieve something where the module-functions return a promise which is resolved by the main class, and where the main class gives that promise a reference to itself (the main class) as the argument to the promise callback (so that the modules would be able to read the data they need while their promises are being resolved). But that design would still not work very well because many of the steps that the functions need to execute are referencing other classes, so it'd be a mess for us to set this up and maintain it.

The only remaining solution I can think of is to make all functions static, but that's a huge mess (look at the below and imagine it for hundreds of functions, with that annoying `Instagram $ig` literally everywhere)

```php
abstract class Timeline
{
    public static function uploadVideo(Instagram $ig, $data) {
        return Internal::uploadVideo($ig, $data);
    }
}

abstract class Internal
{
    public static function uploadVideo(Instagram $ig, $filename) {
        // bla bla
        $result = self::uploadChunks($ig, $filename);
        // bla
        return $result;
    }

    public static function useSegmentedUploader(Instagram $ig) {
        return $ig->checkExperiment('something');
    }

    public static function uploadChunks(Instagram $ig, $filename) {
       $upload = $ig->request('upload/')->addFile($filename);
       // bla bla
       return $result;
    }
}
```

Usage (it gets very ugly and tedious and kills efficient autocomplete since the classnames are not easy to autocomplete, and it's very bug-prone if anyone accidentally doesn't include the $ig parameter)):

```php
\InstagramAPI\Timeline::uploadVideo($ig, $data);
\InstagramAPI\Timeline::getSelfUserFeed($ig);
\InstagramAPI\Media::comment($ig, $media, 'hi');
\InstagramAPI\Live::getTopLive($ig);
\InstagramAPI\Timeline::uploadVideo($data); // oops bug
```

vs:

```php
$ig->timeline->uploadVideo($data);
$ig->timeline->getSelfUserFeed();
$ig->timeline->comment($media, 'hi');
$ig->timeline->getTopLive();
```

The current design is by far better. It's not a God Object. It's just a central settings storage with quick-and-easy references to all necessary modules: In other words, references that make the inter-class communication easy. Thanks to this design, it's very easy to reach the functions within each module. All function signatures are clean. Autocomplete works perfectly. It's great.

The answer to the problem in this ticket is: **PHP** has a bad garbage collector. The design pattern we are using in this library is fine, and work perfectly in any language except PHP. Should we really have to make the class-code a mess just to create a workaround for PHP's shitty garbage collector?

Final thoughts: Most people use 1 instance of the class and that's all they ever use per runtime. Do people really have a legitimate need to create and release lots of instances of the class during a single runtime? Why would they want to do that? I can't think of _any_ reasons. If they want the old class to "die" they can just switch its active user on each "to-kill" instance via `login()` instead and keep reusing that instance without ever killing/destroying it. And if they _do_ want to create lots of instances and kill them, then sure, but then it's _their_ job to call `gc_collect_cycles();` manually, to tell PHP's shitty garbage collector to do a proper job.

**Comment by mgp25** [comment]
@SteveJobzniak nice explanation as always. Good to see you around here! Hope you are enjoying your time 😄

**Comment by valga** [comment]
@SteveJobzniak Spring is late this year, so I can't enjoy it yet.

Well, we can refactor our code to remove a central settings storage aka God Object without breaking everything. Dependency injection comes to rescue:
```php

class Client {
  private $parent;
  public function __construct(Instagram $parent) {
    $this->parent = $parent;
  }
  ...
    $this->parent->settings->setCookies('');
  ...
}

vs

class Client {
  private $settings;
  public function __construct(StorageHandler $settings) {
    $this->settings = $settings;
  }
  ...
    $this->settings->setCookies('');
  ...
}
```
You might say "But what about device? It changes with account switch, so it can't be injected". Yep, and that is another problem with our code: the `Instagram` class is a mix of user and device related settings. But in reality they are not related: you can use multiple accounts on the same phone.

Basically, our `setUser()` approach is broken. Just look at #1759 and here: https://github.com/mgp25/Instagram-API/blob/d13e3a673b4d54c0cf74d730cb9c6df92b570749/src/Instagram.php#L542-L546

**Comment by SteveJobzniak** [comment]
@mgp25 Thanks mate. My days used to feel like this:

![giphy](https://user-images.githubusercontent.com/22373193/37880925-3ebb2fd6-3090-11e8-968b-09ddb892ede3.gif)

Now they feel like this:

![giphy-2](https://user-images.githubusercontent.com/22373193/37880928-4eabd5ee-3090-11e8-876f-b02f9a4c946b.gif)

I am really happy. Things are going well. :-) I hope the same for you!

@valga Hehe... let's back up a bit and remember the big picture... We're talking about 10 kilobytes of RAM... 10 kb... which can _only_ be lost in some _very rare_ misuse-cases (almost no users are affected by this since most people use/reuse a single class instance), and it's _only_ happening because of PHP's very basic garbage collector.

Always ask this question: Is this important enough to care? In this case it's a giant waste of time to refactor the library, because when all of this hard work is done and has consumed many hours or days out of your life, your results would only affect the _one or two persons_ in the world, who have decided, for some strange reason, that they want to constantly create _and destroy_ `new Instagram()` instances instead of reusing a single instance or a pool of instances. And worst of all: Those two people _already have_ plenty of available workarounds as described above (single class reuse or a pool of classes via `setUser()`, or manual forced GC)...

So it would be a bunch of hard, totally pointless work that's only useful for two people who are probably doing something shady anyway (what could the actual use-case be for creating lots of `new Instagram()` instances in a loop and being worried about memory?! spamming via lots of different accounts?!)...

But perhaps more importantly: The proposed refactoring doesn't actually _solve_ PHP's bad garbage collector. It just shifts the problem deeper. The parent class would no longer be a problem for PHP's garbage collector... but... after the changes, there's still a risk of cyclical references _between modules_ (and of course all other cyclical variables). If `Timeline` has a reference to `Internal` and `Internal` has a reference to `Timeline`, then each of them would always have a reference count of at least 1, so they won't get garbage collected. So after all of this work, we have just shifted PHP's shitty garbage collection issue one level deeper and haven't solved the core problem: PHP's bad garbage collector...

And again... we're talking about 10kb being wasted for one or two rare people who don't simply reuse a single class instance. Never forget _that_. ;-)

----

On the other hand, you have brought up some _very_ worthwhile, unrelated projects! :-)

- The current `setUser()` approach, as you mention, was written at a time before "Facebook token logins (logins with just a token, without any username)" existed. But that's a separate problem which has another, even bigger issue: We still _need_ the `->username` property for some things. So we always need their current Instagram username even when a "username-less" Facebook token login is performed. There are some other ways to get their Instagram username (and I think the most reliable way is to read it from the session cookies). But either way, that whole discussion belongs in #1759. And I'm in favor of simple solutions such as storing token logins as `row.username = 'fb-'.$fbToken` in the database (and in `->username` for `setUser()` user switch detection purposes), which would allow most code to stay exactly the same and only needs some minimal `setUser()` changes to be handled properly. And retrieving the _actual_ Instagram account username for a new `->profile_username` property (to be usable as a parameter in requests) can be done via the `client->getCookie()` method.
- Being able to use multiple users per device ID would be a nice bonus. And I suspect that it would help with some feature implementations, such as being able to register a single push notification subscription that covers _all_ of the accounts on that "device id" via a single connection? But just beware that if you decide to decouple the device ID from stuff like uuid, then you're going to land in trouble unless you're very careful. The UUID is tied to the device ID, because the UUID is randomly generated by the device at first login, and an UUID is extremely unique (did you know: _after generating 1 billion UUIDs every second for the next 100 years, the probability of creating just one duplicate would be about 50%_). So you can't just start using the same UUID from "other devices". Instagram will easily detect if UUID "xyz123" was being used from device "abc1" and then suddenly starts being used by device "abc2" instead. So when the device ID changes, then _all_ other things must also change and all sessions, cookies, etc must be replaced. _But_ if the device ID changes _back_ to what it was previously, then the _old device's_ UUID must be restored back to what it was, to avoid creating new UUIDs every time you decide to "swap device" back and forth. Because Instagram bans/limits accounts if you login with too many different UUIDs to an account (I discovered this by accident during UUID testing, and fixing that was actually my first contribution to this library). So decoupling the device ID and UUID is very dangerous work. You'd need some kinda "per-device settings storages", so that you have "one storage database per phone device", to emulate real phones that contain real installations of Instagram (with 1 or more accounts per-phone), including the possibility of someone using the same account from multiple phones. So the settings storage would have to be totally rewritten to a device-centric instead of a user-centric approach. Something like "device abc1 contains data about user1, user2" and "device abc2 contains data about user1, user3". And if user1 is used with abc1, load that user's "abc1"  settings, and if it's used with abc2, load that user's "abc2" settings instead. This must be done carefully and in a reliable and user-friendly way. Most people definitely won't even care about the idea of custom/reusable device IDs and will actually just be _annoyed_ by these changes, because to login they would now have to provide their user/password AND the exact same last-used device ID (for that particular account, so that the class knows what "phone device" to read their account settings, UUID, etc from). And there's also a risk of users misusing the system and generating new devices every time and getting their accounts banned/limited due to having logged in from 100s of "device IDs" to a single account. Or the other way: Misusing by logging into all accounts from the same account and getting the accounts banned for that reason instead (too many accounts on 1 device). Now you see why I decided to simply generate one random device ID per user and keep every user on their own individual device IDs as if there's just 1 account per phone. Because the alternative designs are all super complex. You'd really need to design a _true_ "per-phone/device database" system if you decide to go down that path... so I don't recommend doing it unless you have a _very_ good library-reason to want multiple accounts to share the same device ID (such as the push notification theory).
- Many settings (that really belong in the `StorageHandler`) are currently held as statically saved properties on the `Instagram` class. So all things like `$this->ig->uuid` should definitely be rewritten as `$this->ig->settings->get('uuid')`. It would finally get rid of the annoyingly static, central cache of all important values. But care must be taken here, because some flags etc (such as the important `->isMaybeLoggedIn` would also need to be moved to the `StorageHandler` and handled properly there instead... In fact, _everything_ belonging to the current user would have to be moved to `StorageHandler`...

All of this sounds like very annoying work. I remember why I had to leave to survive. 🤣 

My advice to both of you is: Relax and breathe and only give attention to the most important topics, and don't get caught in minor issues, such as this pointless ticket about how "10kb of RAM _may_ be lost in some ultra-rare cases". ;-)

Take care guys. I wish you all the best of luck. I'm feeling alive at last, and this is about to be one of the best years of my life. And remember... if you need to relax, just look at this... and forget about programming for a while. :-D

![giphy-4](https://user-images.githubusercontent.com/22373193/37881183-49d26638-3094-11e8-9545-dc0434a02aec.gif)

OoooooOooooh... spooooky. 👻

**Comment by valga** [comment]
@SteveJobzniak Well, it's not about saving 10KB of RAM (that can be freed by GC anyway), it's about getting rid of bad architecture. We have a bad influence on other developers, you know.

> We still need the ->username property for some things.

`getLoggedInUser()->getUsername()` has it, so we can save it for later use.

> storing token logins as `row.username = 'fb-'.$fbToken` in the database

Access token is a subject to change, so it can't be used like that.

> being able to register a single push notification subscription that covers all of the accounts on that "device id" via a single connection

I am not interested in that, because realtime connection can't be shared between multiple accounts.

> all things like `$this->ig->uuid` should definitely be rewritten as `$this->ig->settings->get('uuid')`

It's a bad idea. We can just add some `Installation` class that will have all per-phone things like uuid, device_id and phone_id.

> such as the important `->isMaybeLoggedIn`

We can just check whether `sessionid` cookie exists. Moreover, we can do it anytime we want.

**Comment by SteveJobzniak** [comment]
> Well, it's not about saving 10KB of RAM (that can be freed by GC anyway), it's about getting rid of bad architecture. We have a bad influence on other developers, you know.

Sure, that's a valid enough reason, but just remember that _someone_ actually has to do the rewrite... ;-) Aren't there _much_ more important things to do?

And remember that you shouldn't use the constructor to do the dependency injection. Whenever A needs B and B needs A (I think we already have a few collections that cross-reference each other), then you'll have to construct both first, and then run something like `->injectDependencies()` on each other... Because otherwise they'd both need an instance of each other to be able to construct (which is impossible).

>> We still need the ->username property for some things.
> getLoggedInUser()->getUsername() has it, so we can save it for later use.

Nah that won't work reliably. That way we only get the username when they do a full login (not a refresh). So we would never detect them changing their username while they're reusing a session.

There is a cookie where Instagram places the username. I've seen it. It always contains the current profile username. Use that one. And grab it immediately after `/login/` or after successful session resume, and probably also after successful `->account->edit()` whenever the user provides a new username parameter to that function (to change their username via the library)...

> I am not interested in that, because realtime connection can't be shared between multiple accounts.

Okay so is there any other reason to want to rearchitecture the device ID system? Because you saw all the bad effects and dangers and user annoyances it would cause, so there needs to be a great reason to decouple accounts from device IDs. People will be really annoyed if they must manually remember what device ID/`Installation` they associated each account with. They'd need to have their own database for that and would hate this library. No other Instagram 3rd party libraries do it that way.

>> all things like `$this->ig->uuid` should definitely be rewritten as `$this->ig->settings->get('uuid')`
> It's a bad idea. We can just add some `Installation` class that will have all per-phone things like uuid, device_id and phone_id.

I kinda like the idea of an "Instagram installation" class which holds per-phone data. Nice idea.

But why do you think it's a bad idea to always read the values from the settings backend? It solves something that has annoyed you in the past: You wanted to be able to share settings and cookies "live" between runtimes. One of the most important things to achieve that would be to _stop_ centrally caching settings in static variables. And instead always read them from the settings backend.

The backend itself is cached too, as you know. So performance is still high this way. And this change means that if you want to write a "shared settings backend" you are now much closer to being able to do that, since you're sure that all settings are fetched from the storage backend/its cache, instead of being stuck in a bunch of unmanageable static places.

>> such as the important `->isMaybeLoggedIn`
> We can just check whether sessionid cookie exists. Moreover, we can do it anytime we want.

That cookie is not reliable. It can be invalid, expired, or revoked (due to account password change). Do a source code search: `ag isMaybeLoggedIn` and you'll see why that variable is needed. One of the most important places is this in Client.php:

```php
            } catch (LoginRequiredException $e) {
                // Instagram told us that our session is invalid (that we are
                // not logged in). Update our cached "logged in?" state. This
                // ensures that users with various retry-algorithms won't hammer
                // their server. When this flag is false, ALL further attempts
                // at AUTHENTICATED requests will be aborted by our library.
                $this->_parent->isMaybeLoggedIn = false;
```
